server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java

```diff
@@ -333,6 +346,33 @@ public void setHasPendingAsyncFetch() {
         this.hasPendingAsyncFetch = true;
     }
 
+    public long unaccountableSearchableSnapshotSize(RoutingNode node) {
+        try {
+            return unaccountableSearchableSnapshotSizes.computeIfAbsent(node, k -> {
+                long totalSize = 0;
+                // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+                for (ShardRouting shard : node.started()) {
+                    if (clusterInfo == null) {
+                        continue;
+                    }
+                    DiskUsage usage = clusterInfo.getNodeMostAvailableDiskUsages().get(node.nodeId());
+                    ClusterInfo.ReservedSpace reservedSpace = clusterInfo.getReservedSpace(
+                        node.nodeId(),
+                        usage != null ? usage.getPath() : ""
+                    );
+                    if (clusterState.metadata().getIndexSafe(shard.index()).isSearchableSnapshot()
```

Comment:
I wonder if we can check `shard.getExpectedShardSize() > 0` first, then check for `isSearchableSnapshot`, just to optimize a bit. Also, I'd like to defer looking up the `usage` and `reservedSpace` until we have determined that it is a searchable snapshot.

Reply:
Awesome! I've added the `shard.getExpectedShardSize() > 0` check and removed `Math.max(shard.getExpecredShardSize, 0) call